components/core/src/ffi/ir_stream/decoding_methods.cpp

```diff
@@ -0,0 +1,437 @@
+        bool header_match = false;
+        if (ir_buf.size() < cProtocol::MagicNumberLength) {
+            return ErrorCode_InComplete_IR;
+        }
+        if (0 == memcmp(ir_buf.data(), cProtocol::FourByteEncodingMagicNumber, cProtocol::MagicNumberLength)) {
+            is_four_bytes_encoding = true;
+            header_match = true;
+        } else if (0 == memcmp(ir_buf.data(), cProtocol::EightByteEncodingMagicNumber, cProtocol::MagicNumberLength)) {
+            is_four_bytes_encoding = false;
+            header_match = true;
+        }
+        if (false == header_match) {
+            SPDLOG_ERROR("Unrecognized encoding");
+            return ErrorCode_Corrupted_IR;
+        }
+        cursor_pos += cProtocol::MagicNumberLength;
+        return ErrorCode_Success;
+    }
+
+    /**
+     * decodes the first message in the given eight-byte encoding IR stream.
+     * if the IR stream is incomplete, return false.
+     * else, return the ending position of the IR stream.
+     * @param ts_info
+     * @param ir_buf
+     * @param message
+     * @param timestamp
+     * @param ending_pos
+     * @return true on success, false otherwise
+     */
+    template <typename encoded_variable_t>
+    static IR_ErrorCode decode_next_message_general (const std::vector<int8_t>& ir_buf,
+                                                     std::string& message,
+                                                     epoch_time_ms_t& timestamp,
+                                                     size_t& ending_pos) {
+        size_t cursor_pos = ending_pos;
+        encoded_tag_t encoded_tag;
+
+        if (false == try_read_tag(ir_buf, cursor_pos, encoded_tag)) {
+            return ErrorCode_InComplete_IR;
+        }
+
+        if (cProtocol::Eof == encoded_tag) {
+            // TODO: do we want to sanity check if the current tag is the last byte of ir_buf
+            return ErrorCode_End_of_IR;
+        }
+
+        std::vector<encoded_variable_t> encoded_vars;
+        std::string all_dict_var_strings;
+        vector<int32_t> dictionary_var_end_offsets;
+        IR_ErrorCode error_code;
+        bool is_encoded_var;
+        // handle variables
+        while (is_variable_tag<encoded_variable_t>(encoded_tag, is_encoded_var)) {
+            if (is_encoded_var) {
+                encoded_variable_t encoded_variable;
+                if (false == read_data_big_endian(ir_buf, encoded_variable, cursor_pos)) {
+                    return ErrorCode_InComplete_IR;
+                }
+                encoded_vars.push_back(encoded_variable);
+            } else {
+                std::string var_str;
+                error_code = parse_dictionary_var(ir_buf, encoded_tag, var_str, cursor_pos);
+                if (ErrorCode_Success != error_code) {
+                    return error_code;
+                }
+                all_dict_var_strings.append(var_str);
```

Comment:
Parsing the string into var_str and then appending it into all_dict_var_stirngs seems redundant. Can we modify try_read_string implementation to append into the string buffer directly (which should also work if the given string buffer is empty)?